File: apps/desktop/electron/providers/codex/sessions.ts
import * as fsPromises from 'node:fs/promises';
import type { Dirent } from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';

export interface SessionUsageSummary {
  /** Model with the highest token usage across scanned sessions, or null. */
  topModel: string | null;
  tokensByModel: Record<string, number>;
  scannedFiles: number;
}

const MAX_FILES = 100;
const MAX_FILE_BYTES = 6 * 1024 * 1024;
const MAX_SCAN_BYTES = 32 * 1024 * 1024;
const WINDOW_MS = 30 * 24 * 60 * 60 * 1000;
const cache = new Map<string, { at: number; result: SessionUsageSummary }>();

export async function readBoundedFile(file: string, maxBytes: number): Promise<string> {
  const handle = await fsPromises.open(file, 'r');
  const chunks: Buffer[] = [];
  let bytesRead = 0;
  try {
    while (bytesRead < maxBytes) {
      const buffer = Buffer.allocUnsafe(Math.min(64 * 1024, maxBytes - bytesRead));
      const result = await handle.read(buffer, 0, buffer.length, null);
      if (result.bytesRead === 0) break;
      chunks.push(buffer.subarray(0, result.bytesRead));
      bytesRead += result.bytesRead;
    }
  } finally {
    await handle.close();
  }
  return Buffer.concat(chunks, bytesRead).toString('utf8');
}

async function listSessionFiles(home: string): Promise<string[]> {
  const roots = [path.join(home, 'archived_sessions'), path.join(home, 'sessions')];
  const files: string[] = [];
  const cutoff = Date.now() - WINDOW_MS;
  let visited = 0;

  async function walk(dir: string, depth: number): Promise<void> {
    if (depth > 4) return;
    let entries: Dirent[];
    try {
      entries = await fsPromises.readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (++visited > 5000) return;
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(full, depth + 1);
      } else if (entry.isFile() && entry.name.endsWith('.jsonl')) {
        files.push(full);
      }
    }
  }

  await Promise.all(roots.map((root) => walk(root, 0)));

  // Keep only recently touched files, newest first.
  const withMtime: Array<{ file: string; mtime: number }> = [];
  for (const file of files) {
    try {
      const mtime = (await fsPromises.stat(file)).mtimeMs;
      if (mtime >= cutoff) withMtime.push({ file, mtime });
    } catch {
      /* file vanished */
    }
  }
  withMtime.sort((a, b) => b.mtime - a.mtime);
  return withMtime.slice(0, MAX_FILES).map((entry) => entry.file);
}

interface ModelTally {
  tokens: number;
  turns: number;
}

type JsonObject = Record<string, unknown>;

function asObject(value: unknown): JsonObject | null {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? value as JsonObject
    : null;
}

function asTokenCount(value: unknown): number | null {
  return typeof value === 'number' && Number.isFinite(value) && value >= 0 ? value : null;
}

function usageTotal(value: unknown): number | null {
  const usage = asObject(value);
  if (!usage) return null;
  const total = asTokenCount(usage.total_tokens);
  if (total !== null) return total;
  const input = asTokenCount(usage.input_tokens);
  const output = asTokenCount(usage.output_tokens);
  return input === null && output === null ? null : (input ?? 0) + (output ?? 0);
}

function readModel(...sources: Array<JsonObject | null>): string | null {
  for (const source of sources) {
    const model = source?.model_name ?? source?.model;
    if (typeof model === 'string' && model.length > 0) return model;
  }
  return null;
}

/**
 * Attribute tokens per model from native Codex session logs.
 * `turn_context` lines carry the active model; `event_msg`/`token_count` lines
 * carry usage in `payload.info` (total_token_usage or flat token counts).
 */
function tallyContent(content: string, byModel: Map<string, ModelTally>): void {
  let currentModel: string | null = null;
  let previousTotal: number | null = null;
  let incrementalSinceTotal = 0;
  for (const line of content.split('\n')) {
    let event: JsonObject | null;
    try { event = asObject(JSON.parse(line)); } catch { continue; }
    if (!event) continue;
    const payload = asObject(event.payload);
    if (event.type === 'turn_context') {
      currentModel = readModel(payload, event) ?? currentModel;
      if (currentModel) {
        const tally = byModel.get(currentModel) ?? { tokens: 0, turns: 0 };
        tally.turns += 1;
        byModel.set(currentModel, tally);
      }
      continue;
    }

    if (event.type !== 'event_msg' || payload?.type !== 'token_count') continue;
    const info = asObject(payload.info);
    if (!info) continue;

    const cumulativeTotal = usageTotal(info.total_token_usage) ?? asTokenCount(info.total_token_count);
    const incrementalTotal = usageTotal(info.last_token_usage);
    let tokens = 0;
    if (incrementalTotal !== null) {
      tokens = incrementalTotal;
      if (cumulativeTotal === null) incrementalSinceTotal += incrementalTotal;
    } else if (cumulativeTotal !== null) {
      const counterReset = previousTotal !== null && cumulativeTotal < previousTotal;
      const cumulativeDelta = previousTotal === null || counterReset ? cumulativeTotal : cumulativeTotal - previousTotal;
      tokens = counterReset ? cumulativeDelta : Math.max(0, cumulativeDelta - incrementalSinceTotal);
    }
    if (cumulativeTotal !== null) {
      previousTotal = cumulativeTotal;
      incrementalSinceTotal = 0;
    }

    const model = readModel(info, payload, event) ?? currentModel;
    if (!model || tokens <= 0) continue;
    const tally = byModel.get(model) ?? { tokens: 0, turns: 0 };
    tally.tokens += tokens;
    byModel.set(model, tally);
  }
}

/** Scan local Codex session logs for per-model token usage. Best-effort, never throws. */
export async function scanSessionUsage(
  home = process.env.CODEX_HOME || path.join(os.homedir(), '.codex'),
): Promise<SessionUsageSummary> {
  const cached = cache.get(home);
  if (cached && Date.now() - cached.at < 15 * 60 * 1000) return cached.result;
  const byModel = new Map<string, ModelTally>();
  let scannedFiles = 0;
  let scannedBytes = 0;
  try {
    const files = await listSessionFiles(home);
    for (const file of files) {
      try {
        const stat = await fsPromises.stat(file);
        if (stat.size > MAX_FILE_BYTES || scannedBytes + stat.size > MAX_SCAN_BYTES) continue;
        const content = await readBoundedFile(file, Math.min(MAX_FILE_BYTES, MAX_SCAN_BYTES - scannedBytes));
        scannedBytes += Buffer.byteLength(content);
        if (scannedBytes > MAX_SCAN_BYTES) break;
        tallyContent(content, byModel);
        scannedFiles += 1;
      } catch { /* file vanished or is inaccessible */ }
    }
  } catch {
    /* scanning is best-effort */
  }

  let topModel: string | null = null;
  for (const [model, tally] of byModel) {
    if (tally.tokens <= 0) continue;
    if (!topModel || tally.tokens > (byModel.get(topModel)?.tokens ?? 0)) topModel = model;
  }

  const tokensByModel: Record<string, number> = {};
  for (const [model, tally] of byModel) tokensByModel[model] = tally.tokens;
  const result = { topModel, tokensByModel, scannedFiles };
  if (cache.size >= 21) {
    const oldest = cache.keys().next().value;
    if (oldest) cache.delete(oldest);
  }
  cache.set(home, { at: Date.now(), result });
  return result;
}

File: src-tauri/src/providers/sessions.rs
use serde_json::Value;
use std::collections::{BTreeMap, HashMap};
use std::fs::{self, File};
use std::io::{Read, BufRead, BufReader};
use std::path::{Path, PathBuf};
use std::sync::{Mutex, OnceLock};

const MAX_FILES: usize = 100;
const MAX_FILE_BYTES: u64 = 6 * 1024 * 1024;
const MAX_SCAN_BYTES: u64 = 32 * 1024 * 1024;
const WINDOW_MS: i64 = 30 * 24 * 60 * 60 * 1000;
const CACHE_TTL_MS: i64 = 15 * 60 * 1000;

#[derive(Debug, Clone, Default)]
pub struct SessionUsageSummary {
    pub top_model: Option<String>,
}

static CACHE: OnceLock<Mutex<HashMap<PathBuf, (i64, SessionUsageSummary)>>> = OnceLock::new();

#[derive(Default)]
struct ModelTally {
    tokens: u64,
    turns: u64,
}

pub fn scan_session_usage(home: &Path) -> SessionUsageSummary {
    let cache = CACHE.get_or_init(|| Mutex::new(HashMap::new()));
    let now = crate::forecast::now_millis();
    if let Ok(cache) = cache.lock() {
        if let Some((at, summary)) = cache.get(home).filter(|(at, _)| now - *at < CACHE_TTL_MS) {
            let _ = at;
            return summary.clone();
        }
    }

    let mut tallies: BTreeMap<String, ModelTally> = BTreeMap::new();
    let mut scanned_bytes = 0_u64;
    for file in list_session_files(home, now) {
        let Ok(metadata) = fs::metadata(&file) else { continue };
        if metadata.len() > MAX_FILE_BYTES || scanned_bytes + metadata.len() > MAX_SCAN_BYTES { continue }
        let Ok(content) = read_bounded_file(&file, MAX_FILE_BYTES.min(MAX_SCAN_BYTES - scanned_bytes)) else { continue };
        scanned_bytes += content.len() as u64;
        if scanned_bytes > MAX_SCAN_BYTES { break }
        tally_content(&content, &mut tallies);
    }

    let top_model = tallies.iter().filter(|(_, tally)| tally.tokens > 0).max_by_key(|(_, tally)| tally.tokens).map(|(model, _)| model.clone());
    let summary = SessionUsageSummary {
        top_model,
    };
    if let Ok(mut cache) = cache.lock() {
        if cache.len() >= 21 {
            if let Some(oldest) = cache.iter().min_by_key(|(_, (at, _))| *at).map(|(key, _)| key.clone()) {
                cache.remove(&oldest);
            }
        }
        cache.insert(home.to_path_buf(), (now, summary.clone()));
    }
    summary
}

fn read_bounded_file(file: &Path, max_bytes: u64) -> std::io::Result<String> {
    let handle = File::open(file)?;
    let mut content = String::new();
    handle.take(max_bytes).read_to_string(&mut content)?;
    Ok(content)
}

fn list_session_files(home: &Path, now: i64) -> Vec<PathBuf> {
    let cutoff = now.saturating_sub(WINDOW_MS);
    let mut visited = 0_usize;
    let mut candidates = Vec::new();
    for root in [home.join("archived_sessions"), home.join("sessions")] {
        for entry in walkdir::WalkDir::new(root).follow_links(false).max_depth(5).into_iter().filter_map(Result::ok) {
            visited += 1;
            if visited > 5000 { break }
            if !entry.file_type().is_file() || !entry.file_name().to_string_lossy().ends_with(".jsonl") { continue }
            let Ok(metadata) = entry.metadata() else { continue };
            let modified = metadata.modified().ok().and_then(|time| time.duration_since(std::time::UNIX_EPOCH).ok()).map(|duration| duration.as_millis() as i64).unwrap_or(0);
            if modified >= cutoff { candidates.push((entry.path().to_path_buf(), modified)); }
        }
    }
    candidates.sort_by(|a, b| b.1.cmp(&a.1));
    candidates.into_iter().take(MAX_FILES).map(|(path, _)| path).collect()
}

fn object(value: Option<&Value>) -> Option<&serde_json::Map<String, Value>> {
    value?.as_object()
}

fn token_count(value: Option<&Value>) -> Option<u64> {
    value?.as_u64()
}

fn usage_total(value: Option<&Value>) -> Option<u64> {
    let usage = object(value)?;
    if let Some(total) = token_count(usage.get("total_tokens")) { return Some(total) }
    let input = token_count(usage.get("input_tokens"));
    let output = token_count(usage.get("output_tokens"));
    if input.is_none() && output.is_none() { None } else { Some(input.unwrap_or(0).saturating_add(output.unwrap_or(0))) }
}

fn read_model(sources: &[Option<&Value>]) -> Option<String> {
    for source in sources {
        let Some(map) = object(*source) else { continue };
        let model = map.get("model_name").or_else(|| map.get("model")).and_then(Value::as_str);
        if let Some(model) = model.filter(|model| !model.is_empty()) { return Some(model.to_string()) }
    }
    None
}

fn tally_content(content: &str, tallies: &mut BTreeMap<String, ModelTally>) {
    let mut current_model: Option<String> = None;
    let mut previous_total: Option<u64> = None;
    let mut incremental_since_total = 0_u64;
    for line in BufReader::new(content.as_bytes()).lines().map_while(Result::ok) {
        let Ok(event) = serde_json::from_str::<Value>(&line) else { continue };
        let payload = event.get("payload");
        if event.get("type").and_then(Value::as_str) == Some("turn_context") {
            current_model = read_model(&[payload, Some(&event)]).or(current_model);
            if let Some(model) = &current_model { tallies.entry(model.clone()).or_default().turns += 1; }
            continue;
        }
        if event.get("type").and_then(Value::as_str) != Some("event_msg") || payload.and_then(|value| value.get("type")).and_then(Value::as_str) != Some("token_count") { continue }
        let Some(info) = payload.and_then(|value| value.get("info")) else { continue };
        let cumulative = usage_total(info.get("total_token_usage")).or_else(|| token_count(info.get("total_token_count")));
        let incremental = usage_total(info.get("last_token_usage"));
        let tokens = if let Some(incremental) = incremental {
            if cumulative.is_none() { incremental_since_total = incremental_since_total.saturating_add(incremental); }
            incremental
        } else if let Some(cumulative) = cumulative {
            let reset = previous_total.is_some_and(|previous| cumulative < previous);
            let delta = if previous_total.is_none() || reset { cumulative } else { cumulative.saturating_sub(previous_total.unwrap_or(0)) };
            if reset { delta } else { delta.saturating_sub(incremental_since_total) }
        } else { 0 };
        if let Some(cumulative) = cumulative {
            previous_total = Some(cumulative);
            incremental_since_total = 0;
        }
        let model = read_model(&[Some(info), payload, Some(&event)]).or_else(|| current_model.clone());
        if let Some(model) = model.filter(|_| tokens > 0) {
            let tally = tallies.entry(model).or_default();
            tally.tokens = tally.tokens.saturating_add(tokens);
        }
    }
}
